projects/astc-encoder/astc_load_decompress_image_fuzzer.cc

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "astcenc_internal.h"
+#include "astcenccli_internal.h"
+
+#include <fuzzer/FuzzedDataProvider.h>
+#include <limits>
+#include <vector>
+
+static constexpr uint8_t kUint8Max = std::numeric_limits<uint8_t>::max();
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+
+  if (size < 19)
+    return 0;
+
+  astc_compressed_image image_comp;
+  astcenc_profile profile;
+  int out_bitness;
+  FuzzedDataProvider stream(data, size);
+
+  const astcenc_profile profiles[] = {ASTCENC_PRF_LDR, ASTCENC_PRF_LDR_SRGB,
+                                      ASTCENC_PRF_HDR,
+                                      ASTCENC_PRF_HDR_RGB_LDR_A};
+
+  int profile_type = stream.ConsumeIntegralInRange<int>(0, 3);
+  profile = profiles[profile_type];
+  out_bitness = 8 << (profile_type / 2);
+
+  // Avoid dividing by zero
+  uint8_t block_x = stream.ConsumeIntegralInRange<uint8_t>(1, kUint8Max);
+  uint8_t block_y = stream.ConsumeIntegralInRange<uint8_t>(1, kUint8Max);
+  uint8_t block_z = stream.ConsumeIntegralInRange<uint8_t>(1, kUint8Max);
+
+  // Reference file consumes 3 bytes for each, so we define a maximum value
+  unsigned int dim_x =
+      stream.ConsumeIntegralInRange<uint32_t>(1, ~(0xff << 24));
+  unsigned int dim_y =
+      stream.ConsumeIntegralInRange<uint32_t>(1, ~(0xff << 24));
+  unsigned int dim_z =
+      stream.ConsumeIntegralInRange<uint32_t>(1, ~(0xff << 24));
+
+  unsigned int xblocks = (dim_x + block_x - 1) / block_x;
+  unsigned int yblocks = (dim_y + block_y - 1) / block_y;
+  unsigned int zblocks = (dim_z + block_z - 1) / block_z;
+
+  // Following the structure of
+  // ARM-software/astc-encoder/Source/astcenccli_toplevel.cpp:main(), located at
+  // https://github.com/ARM-software/astc-encoder/blob/master/Source/astcenccli_toplevel.cpp
+  size_t buffer_size = xblocks * yblocks * zblocks * 16;
+  if (size - 19 < buffer_size)
```

Comment:
this check again depends on the implementation of FDP, it might have consumed more or less than 19 bytes...
either check if at least buffer_size is remaining or consume the buffer size and check that it has the required size.

Reply:
if you even don't check at all, if enough data was left for consumption, we might find issues where the library accesses data even if we tell it how much data there is to access, i think that would be better...